_posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

```diff
@@ -0,0 +1,182 @@
+---
+layout: post
+title:  "Enhanced Multi Vector Support in OpenSearch KNN with Nested Field"
+authors:
+- heemin
+- vamshin
+- dylantong
+date: 2024-02-23 00:00:00 -0700
+categories:
+- technical-posts
+meta_keywords: OpenSearch multi vector, OpenSearch knn nested field
+meta_description: Improvement in OpenSearch KNN with Nested Field, specifically focusing on multi-vector support
+excerpt: In OpenSearch 2.12, users can now obtain diverse search results from the KNN index, even when multiple nearest vectors belong to just a few documents. This enhancement is expected to enhance both the efficiency and quality of search outcomes.
+---
+
+OpenSearch 2.12 brings a significant enhancement to its KNN index with HNSW algorithm in Faiss and Lucene engine, boosting result diversity by intelligently handling multi-vector support. This improvement ensures better search outcomes by effectively eliminating duplicate vectors from the same document during KNN searches.
+
+## Why is this important?
+With the rise of Large Language Models (LLM), many users are turning to vector databases for indexing, storing, and retrieving information, particularly for building Retrieval-augmented Generation (RAG) systems. These models convert text to embeddings and preserve the semantic information in the embeddings. However, models have limitations on number of tokens to consider for embedding generation which requires document to be chunked and stored as multiple embeddings in a single document. Without multi-vector support, every chunk of the document needs to be treated as a separate entity, requiring duplication of document metadata across multiple chunk unless the user separates the document and its metadata. This duplication of metadata can result in significant storage and memory usage. Furthermore, users must independently devise a merging mechanism to ensure that only a single document is retrieved from multiple chunk during search. Multi-vector support simplifies the handling of large documents, alleviating these challenges.
+
+## Understanding previous limitations
+### Fewer Results
+In previous versions of OpenSearch, when searching for nested KNN (K-Nearest Neighbors) fields with a specified number of nearest neighbors (k value), the search might return fewer than k documents. This occurred because the search operated at the nested field level or chunk level rather than the document level.  In the worst scenario, its possible that k neighbors could be the chunks belonging to the single document, resulting in just one document in search results than desired. While increasing the k value could retrieve more documents, it introduced unnecessary search overhead.
```

Comment:
```suggestion
In versions of OpenSearch prior to 2.12, when you search for nested k-NN fields by providing a specified number of nearest neighbors (k value), the search might return fewer than k documents. In those OpenSearch versions, the search operates at the nested field level or chunk level rather than the document level.  In the worst case scenario, the retrieved k neighbors can be the chunks belonging to a single document, resulting in just one document in search results. While increasing the k value can retrieve more documents, it introduces unnecessary search overhead.
```